content/6_Advanced/Eulers_Formula.mdx

```diff
@@ -1,104 +1,152 @@
+
+	int find(int u) {
+		if (parent[u] != u) { parent[u] = find(parent[u]); }
+		return parent[u];
+	}
+
+	void unionSets(int u, int v) {
+		int rootU = find(u);
+		int rootV = find(v);
+		if (rootU != rootV) {
+			if (rank[rootU] > rank[rootV]) {
+				parent[rootV] = rootU;
+			} else if (rank[rootU] < rank[rootV]) {
+				parent[rootU] = rootV;
+			} else {
+				parent[rootV] = rootU;
+				rank[rootU]++;
 			}
+			components--;
 		}
-		comp[x].pb(hsh(t.f, t.s));
 	}
-	void unite(pi x, pi y) {  // union-by-rank
-		int X = hsh(x.f, x.s), Y = hsh(y.f, y.s);
-		if (par[X] == par[Y]) return;
-		X = par[X], Y = par[Y];
-		if (sz(comp[X]) < sz(comp[Y])) swap(X, Y);
-		trav(t, comp[Y]) addPoint(X, {t / N, t % N});
-		comp[Y].clear();
+
+	void addComponent(int u) {
+		if (parent[u] == u) { components++; }
 	}
+
+	int getComponentsCount() const { return components; }
 };
 
-DSU<750 * 750> D;
-bool ok[750][750];
+int main() {
+	int n, m;
+	cin >> n >> m;
+
+	vector<vector<char>> grid(n, vector<char>(n, '.'));  // '.' for water
+	UnionFind uf(n * n);
+	vector<bool> hasLake(n * n, false);
+
+	auto index = [&](int r, int c) { return r * n + c; };
+
+	auto isValid = [&](int r, int c) {
+		return r >= 0 && r < n && c >= 0 && c < n;
+	};
+
+	char type;
+	int r, c;
+
+	for (int i = 0; i < m; ++i) {
+		cin >> type;
+		if (type == '!') {
+			cin >> r >> c;
+			r--;
+			c--;
+
+			if (grid[r][c] == '#') {
+				grid[r][c] = '.';                        // Change land to water
+				uf.unionSets(index(r, c), index(r, c));  // Handle disconnecting
+			} else {
+				grid[r][c] = '#';  // Change water to land
+				uf.addComponent(index(r, c));
+
+				// Check adjacent cells
+				static const vector<pair<int, int>> directions = {
+				    {-1, 0}, {1, 0}, {0, -1}, {0, 1}};
+				for (const auto &dir : directions) {
+					int nr = r + dir.first;
+					int nc = c + dir.second;
+					if (isValid(nr, nc) && grid[nr][nc] == '#') {
+						uf.unionSets(index(r, c), index(nr, nc));
+					}
+				}
+			}
+		} else if (type == '?') {
+			int islands = uf.getComponentsCount();
+			int noLakeIslands = islands;  // Initialize to total islands
 
-void solve(int x, int y) {
-	ok[x][y] = 1;
-	F0R(i, 4) {
-		int X = x + xd[i], Y = y + yd[i];
-		if (X < 0 || X >= N || Y < 0 || Y >= N) continue;
-		if (!ok[X][Y]) continue;
-		D.unite({x, y}, {X, Y});
-	}
-}
+			// Check for lakes (this needs an extra logic, e.g., flood fill)
+			// This part needs implementation to correctly detect lakes.
 
-int main() {
-	setIO("valleys");
-	re(N);
-	F0R(i, N) F0R(j, N) {
-		re(h[i][j]);
-		v.pb({h[i][j], {i, j}});
-		D.comp[hsh(i, j)].pb(hsh(i, j));
-	}
-	sort(all(v));
-	F0R(i, sz(v)) {
-		solve(v[i].s.f, v[i].s.s);
-		pi p = v[i].s;
-		int q = D.par[hsh(p.f, p.s)];
-		if (D.measure[q] == 1) ans += sz(D.comp[q]);
+			cout << islands << " " << noLakeIslands << endl;
+		}
 	}
-	cout << ans;
+
+	return 0;
 }
 ```
 
+</CPPSection>
+</LanguageSection>
+
+## Valleys
+
+<Problems problems="valleys" />
```

Comment:
why is the problem header named valleys

Reply:
Ohhhh I see. When you said "add a focus problem" on the [usaco-950.mdx](https://github.com/cpinitiative/usaco-guide/pull/4601/files/059f76400c9d19b4add38b1588257f9d89dd773f..40f48aa46861cf3fe466087f32267ec9ceee037a#diff-7b9c8da0f0c2ae020c1d0b48a9deaf8360068ed52feb25cafc6b4823db86819c) file, I thought you wanted me to add it as a focus problem to the Euler's module so I gave it a problem header named Valleys but was unsure whether to add explanation since it already had an internal solution with all of that.

Reply:
oh yeah it still deserves a module explanation